src/computational_graph/graph.jl

```diff
@@ -255,124 +282,212 @@ end
 
 julia> g.subgraphs
-3-element Vector{Graph{Float64, Float64}}:
-1:f⁻(5)|f⁺(1)=0.0
-2:f⁻(2)|f⁺(4)=0.0
-3:ϕ(3)|ϕ(6)=0.0
+6-element Vector{Graph{Float64, Float64}}:
+ 1:f⁺(1)f⁻(2)ϕ(3)=0.0
+ 2:f⁺(4)f⁻(5)ϕ(6)=0.0
+ 3:f⁺(7)f⁻(8)ϕ(9)=0.0
+ 4:f⁺(1)|f⁻(5)⋅-1.0=0.0
+ 5:ϕ(3)|ϕ(9)=0.0
+ 6:f⁺(4)|f⁻(8)⋅-1.0=0.0
 ```
 """
-function feynman_diagram(vertices::AbstractVector,
-    topology::Vector{Vector{Int}};
-    external=[], factor=one(_dtype.factor),
-    weight=zero(_dtype.weight),
-    name="", type=:generic)
-
-    # convert all Graphs to vertices by grouping all its external legs as a vertex
-    converted_vertices = [_extract_vertex(typeof(v), v) for v in vertices]
-    # collect graphs in vertices as subdiagrams
-    subdiagrams = [v for v in vertices if isa(v, Graph)]
-
-    operators = [o for v in converted_vertices for o in v.operators]
-    permutation = collect(Iterators.flatten(topology))
-    ind_ops = collect(eachindex(operators))
-
-
-    @assert length(unique(permutation)) == length(permutation) # no repeated index
-    @assert length(unique(external)) == length(external) # no repeated index
-    # @assert Set(permutation) == Set(ind_ops) # permutation must exhaust all operators
-    ind_ghost = filter(p -> isghost(operators[p]), ind_ops)
-    @assert all(ind_ghost .<= length(external)) # external real/fake legs must be placed at the beginning of vertices.
-
-    ind_fakeleg = Int[]
-    subgraphs = Graph[]
-    for connection in topology
-        if isempty(intersect(connection, ind_ghost))
-            push!(subgraphs, propagator(operators[connection]))
+function feynman_diagram(subgraphs::Vector{Graph{F,W}}, topology::Vector{Vector{Int}}; perm_noleg::Union{Vector{Int},Nothing}=nothing,
+    factor=one(_dtype.factor), weight=zero(_dtype.weight), name="", diagtype::GraphType=GenericDiag()) where {F,W}
+
+    # external_ops = OperatorProduct(operators[external]) # the external operators for the building diagram after contractions
+    contraction = collect(Iterators.flatten(topology))
+    @assert length(unique(contraction)) == length(contraction)  # no repeated index
+
+    vertices, all_hasLeg = OperatorProduct[], Bool[]
+    external_leg, external_noleg = Int[], Int[] # index all leg/nonleg external operators
+    ind = 0
+    for g in subgraphs
+        g.type == Propagator && continue  # exclude propagator subgraph to avoid double counting.
+        push!(vertices, external(g))
+        append!(all_hasLeg, g.hasLeg)
+        if g.type == ExternalVertex
+            append!(external_leg, g.external .+ ind) # ExternalVertex will be legged after contraction.
         else
-            @assert length(connection) == 2 "Ghost external operator can only be connected to a single internal operator"
-            ind_fop = setdiff(connection, ind_ghost)
-            append!(ind_fakeleg, ind_fop)
+            gext = setdiff(g.external .+ ind, contraction)
+            gextLeg = g.hasLeg[gext.-ind]
+            append!(external_leg, gext[gextLeg])
```

Comment:
add comment: select gext[i] so that gextLeg[i]=true. The selected gext[i] are the external vertices with legs